[DOC] debug backends (#26669)
matplotlib/matplotlib: galleries/users_explain/figure/backends.rst

What is a backend?
------------------

> Backends are used for out-putting Matplotlib figures (see :ref:`figure-intro`),
> to the screen or for writing to files.  A lot of documentation on the website

Reviewer: ```suggestion
Backends are used for displaying Matplotlib figures (see :ref:`figure-intro`)
on the screen, or for writing to files.  A lot of documentation on the website
```

> on the screen, or for writing to files.  A lot of documentation on the website

Reviewer: ```suggestion
on the screen, or for writing to files. A lot of documentation on the website
```

Author: Is this really a thing?  Old people need two spaces.

Reviewer: But you removed the two spaces in the rest of the paragraph?

and in the mailing lists refers to the "backend" and many new users are
confused by this term. Matplotlib targets many different use cases and output
formats.  Some people use Matplotlib interactively from the Python shell and
have plotting windows pop up when they type commands. Some people run `Jupyter
<https://jupyter.org>`_ notebooks and draw inline plots for quick data
analysis. Others embed Matplotlib into graphical user interfaces like PyQt or

> PyGObject to build rich applications.  Some people use Matplotlib in batch

Author: fixed

``matplotlib.use('module://name.of.the.backend')``.

Information for backend implementers is available at :ref:`writing_backend_interface`.

.. _figures_not_showing:

Debugging the figure windows not showing
----------------------------------------

Sometimes things do not work as expected, usually during an install.

If you are using a Notebook or integrated development environment (see :ref:`notebooks-and-ides`),
please consult their documentation for debugging figures not working in their
environments.

If you are using one of Matplotlib's graphics backends (see :ref:`standalone-scripts-and-interactive-use`), make sure you know which
one is being used:

.. code-block:: python3

   import matplotlib

   print(matplotlib.get_backend())

> Try a simple plot, as see if the GUI opens:

Reviewer: ```suggestion
Try a simple plot, to see if the GUI opens:
```

.. code-block:: python3

   import matplotlib
   import matplotlib.pyplot as plt

   print(matplotlib.get_backend())

>    plt.plot((1,4,6))

Reviewer: ```suggestion
   plt.plot((1, 4, 6))
```

   plt.show()

If it does not, you perhaps have an installation problem.  A good step at this
point is to ensure that your GUI toolkit is installed properly, taking

> Matplotlib our of the testing.  Almost all GUI toolkits have a small test

Reviewer: ```suggestion
Matplotlib out of the testing.  Almost all GUI toolkits have a small test
```

program that can be run to test basic functionality.  If this test fails, try re-installing.

> PyQt(5,6) and PySide

Reviewer: I'm not sure about these heading titles, as previously we've said to look at `plt.get_backend()`, which does not return _any_ of these titles.

Author: Do we need both the backend name and the gui module name?

Reviewer: I think I would use the backend names as titles, and then in the text explain that they use xyz GUI module names.

Author: OK, done.

^^^^^^^^^^^^^^^^^^^^

.. code-block:: bash

   python -c "from PyQt5.QtWidgets import *; app = QApplication([]); win = QMainWindow(); win.show(); app.exec()"

If you have ``PySide`` or ``PyQt6`` installed instead, just change the import
accordingly.

tkinter
^^^^^^^

.. code-block:: bash

   python3 -c "from tkinter import Tk; Tk().mainloop()"

GTK
^^^

.. code-block:: bash

>    python3 -c "from gi.repository import Gtk; win = Gtk.Window(); win.connect('destroy', Gtk.main_quit); win.show(); Gtk.main()""

Reviewer: ```suggestion
   python3 -c "from gi.repository import Gtk; win = Gtk.Window(); win.connect('destroy', Gtk.main_quit); win.show(); Gtk.main()"
```

please consult their documentation for debugging figures not working in their
environments.

If you are using one of Matplotlib's graphics backends (see :ref:`standalone-scripts-and-interactive-use`), make sure you know which
one is being used:

.. code-block:: python3

   import matplotlib

   print(matplotlib.get_backend())

Try a simple plot to see if the GUI opens:

.. code-block:: python3

   import matplotlib
   import matplotlib.pyplot as plt

   print(matplotlib.get_backend())
   plt.plot((1, 4, 6))
   plt.show()

If it does not, you perhaps have an installation problem.  A good step at this
point is to ensure that your GUI toolkit is installed properly, taking
Matplotlib out of the testing.  Almost all GUI toolkits have a small test
program that can be run to test basic functionality.  If this test fails, try re-installing.

QtAgg
^^^^^

Test ``PyQt5``:

.. code-block:: bash

   python -c "from PyQt5.QtWidgets import *; app = QApplication([]); win = QMainWindow(); win.show(); app.exec()"

If you have ``PySide`` or ``PyQt6`` installed rather than ``PyQt5``, just change the import

> accordingly.

Reviewer: This maybe needs to be explained before the code.

TkAgg

> ^^^^^

Reviewer: Also TkCairo.

Test ``tkinter``:

.. code-block:: bash

   python3 -c "from tkinter import Tk; Tk().mainloop()"

GTK3Agg, GTK4Agg, GTK3Cairo, GTK4Cairo
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Test ``Gtk``:

.. code-block:: bash

   python3 -c "from gi.repository import Gtk; win = Gtk.Window(); win.connect('destroy', Gtk.main_quit); win.show(); Gtk.main()"

wxAgg

> ^^^^^

Reviewer: Also wxCairo.

``matplotlib.use('module://name.of.the.backend')``.

Information for backend implementers is available at :ref:`writing_backend_interface`.

> .. _figures_not_showing:

Reviewer: We seem to use dashes usually?
```suggestion
.. _figures-not-showing:
```

Sometimes things do not work as expected, usually during an install.

If you are using a Notebook or integrated development environment (see :ref:`notebooks-and-ides`),
please consult their documentation for debugging figures not working in their
environments.

If you are using one of Matplotlib's graphics backends (see :ref:`standalone-scripts-and-interactive-use`), make sure you know which
one is being used:

.. code-block:: python3

   import matplotlib

   print(matplotlib.get_backend())

Try a simple plot to see if the GUI opens:

.. code-block:: python3

   import matplotlib
   import matplotlib.pyplot as plt

   print(matplotlib.get_backend())
   plt.plot((1, 4, 6))
   plt.show()

If it does not, you perhaps have an installation problem.  A good step at this
point is to ensure that your GUI toolkit is installed properly, taking
Matplotlib out of the testing.  Almost all GUI toolkits have a small test
program that can be run to test basic functionality.  If this test fails, try re-installing.

QtAgg, QtCairo, Qt5Agg, and Qt5Cairo
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Test ``PyQt5``:

If you have ``PySide`` or ``PyQt6`` installed rather than ``PyQt5``, just change the import

> accordingly:

Reviewer: Doubled colon in the paragraphs.

Author: Sorry, not sure what you mean here....

Reviewer: Both paragraphs end in a colon, implying that something should come after, but there's nothing supposed to be between these two paragraphs. Perhaps just leave out the first paragraph?

Author: Oh duh, I see now.  Thanks,

Author: ... fixed (finally)

.. _figures_not_showing:

Debugging the figure windows not showing
----------------------------------------

Sometimes things do not work as expected, usually during an install.

If you are using a Notebook or integrated development environment (see :ref:`notebooks-and-ides`),
please consult their documentation for debugging figures not working in their
environments.

If you are using one of Matplotlib's graphics backends (see :ref:`standalone-scripts-and-interactive-use`), make sure you know which
one is being used:

.. code-block:: python3

   import matplotlib

   print(matplotlib.get_backend())

Try a simple plot, as see if the GUI opens:

.. code-block:: python3

   import matplotlib
   import matplotlib.pyplot as plt

   print(matplotlib.get_backend())
   plt.plot((1,4,6))
   plt.show()

If it does not, you perhaps have an installation problem.  A good step at this
point is to ensure that your GUI toolkit is installed properly, taking
Matplotlib our of the testing.  Almost all GUI toolkits have a small test
program that can be run to test basic functionality.  If this test fails, try re-installing.

> PyQT

Reviewer 2: import name is PyQt(5,6) with a lowercase t (throughout).

^^^^

.. code-block:: bash

   python -c "from PyQt5.QtWidgets import *; app = QApplication([]); win = QMainWindow(); win.show(); app.exec()"

If you have ``PySide`` or ``PyQT6`` installed instead, just change the import
accordingly

GTK
^^^

.. code-block:: bash

>    python3 -c 'from gi.repository import Gtk; win = Gtk.Window(); win.connect("destroy", Gtk.main_quit); win.show(); Gtk.main()'

Reviewer 2: Should use double quotes outside and single quotes inside (rather than the other way round), because that'll also work on Windows.
